### PR TITLE
Vendor i18n gem for future internationali[sz]ation work

### PR DESCRIPTION
**Pull request by L2G**
On the mailing list, I recently [announced](http://librelist.com/browser/homebrew/2015/4/23/experimental-internationalized-homebrew-i18nbrew/) my experimental fork of Homebrew with internationalization, [i18nbrew](https://github.com/L2G/i18nbrew).

@mikemcquaid suggested that I go ahead and submit a pull request with minimal changes required to enable i18n, while stopping short of introducing any translations. So this pull request vendors the [i18n gem](https://rubygems.org/gems/i18n) and enables its use in Homebrew.

I made sure that the tests in Homebrew/Library/test still pass.

### CODE REVIEW

**Comment by mistydemeo** [comment]
It looks like this requires Ruby 1.9+ - do you think it's feasible to maintain an i18n-free fallback for users running OS X 10.8 and lower?

**Comment by L2G** [comment]
What's requiring 1.9, the i18n gem? There might be an older version that doesn't require 1.9.

**Comment by L2G** [comment]
It might be possible to vendor i18n 0.6.11 instead of 0.7.0; that looks like the newest version that still supports Ruby 1.8.

**Comment by L2G** [comment]
i18n 0.6.11 seems to work fine, going by my rudimentary tests. The only problem is an annoying deprecation message I haven't figured out how to silence yet, but that's small potatoes.

**Comment by mistydemeo** [comment]
Thanks! If possible, I'd still like to have a fallback that allows Homebrew to work without loading this; I have my doubts it'll work on 1.8.7 (OS X 10.5) or 1.8.2 (OS X 10.4).

**Comment by L2G** [comment]
It would be more practical for me to test the heck out of the i18n gem against OS X 10.4 and 10.5, and patch it for bugs that pop up, than it would be to write a fallback that would essentially be rewriting the i18n gem…

Is there a VM with OS X 10.4 and 10.5 that I could use for testing? Or if I can simply prove that the i18n gem's tests pass against Ruby 1.8.2 and Ruby 1.8.7, would that quell your concerns?

**Comment by jacknagel** [comment]
Ultimately, we probably don't need to vendor all the gemfiles or the i18n gem's own tests, right? Would like to keep the footprint as small as possible.

**Comment by L2G** [comment]
The gemfiles, maybe not. The tests would probably be good to keep if we want to verify that it's working under 10.4 and 10.5.

**Comment by xu-cheng** [comment]
:+1: for internationalization works. For what is worth, we can internationalize [Homebrew/homebrew-bundle](https://github.com/Homebrew/homebrew-bundle) first to see some feedback (considering it has a smaller codebase).

File: .gitignore [hunk 1]
```diff
@@ -10,6 +10,7 @@
 !/bin
 /bin/*
 !/bin/brew
+!/locales
```
Comment (xu-cheng) [inline .gitignore]: Personally, I think it's better to put `locales` inside `Library` as it should be part of Homebrew Library.

Comment (mikemcquaid) [inline .gitignore]: Agreed.

File: LICENSE.txt [hunk 2]
```diff
@@ -20,3 +27,52 @@ DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
...
+The above copyright notice and this permission notice shall be
+included in all copies or substantial portions of the Software.
+
+THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
+LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
+WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
```
Comment (mikemcquaid) [inline LICENSE.txt]: Might want to add any other gems if we're going to do this too CC @jacknagel

File: Library/Homebrew/vendor/i18n.rb [hunk 3]
```diff
@@ -0,0 +1,58 @@
...
+      end
+    end
+  end
+end
+
+::I18n.default_locale = Homebrew::I18n::locale_from_env || :'en-US'
+::I18n.enforce_available_locales = false
+
+def t(*args); I18n.t(*args); end
+def l(*args); I18n.l(*args); end
```
Comment (mikemcquaid) [inline Library/Homebrew/vendor/i18n.rb]: Is this the way you have to configure this gem? Feels a little weird.

**Comment by mikemcquaid** [comment]
I'd agree with deleting as many vendored files as possible. Also, we could just have the `t` and `l` methods just output the strings as-is on Ruby <1.9.

File: LICENSE.txt (hunk 2 above)
Comment (L2G) [inline LICENSE.txt]: I did add okjson's license. That is the only other vendored gem I noticed.

**Comment by L2G** [comment]
The strings being fed to the `t` method are lookup keys like "cmd.installed.already_installed", not actual text messages.

I've been thinking about it a little more, though, and it might be feasible to write a basic shim for i18n that's focused just on English output. But what would determine whether the i18n gem should be switched out in favor of the shim?

File: Library/Homebrew/vendor/i18n.rb (hunk 3 above)
Comment (L2G) [inline Library/Homebrew/vendor/i18n.rb]: What are your concerns?

File: .gitignore (hunk 1 above)
Comment (L2G) [inline .gitignore]: Move the `locales` directory to `Library/Homebrew/locales`, then?

**Comment by L2G** [comment]
A little more detail on the `t` method, just to give you some context:

The first argument to `t` is a lookup key, functionally like a hash key. Additional, optional arguments are in a hash, and provide values to be interpolated in the message looked up via the key.

Example: The English message with the key "cmd.install.devel_only_formula" is:

    %{name} is a devel-only formula
    Install with `brew install --devel %{name}`

So calling `t("cmd.install.devel_only_formula", :name => "fnord")` looks up "cmd.install.devel_only_formula" in the desired locale (English, in this case), interpolates the value of `:name`, and returns the full message:

    fnord is a devel-only formula
    Install with `brew install --devel fnord`

There is a special hash value, `:count`, that invokes some magic related to pluralization. For example, calling `t("utils.minutes", :count => 33)` looks up "utils.minutes" in the desired locale, then chooses a subkey of "utils.minutes" based on that locale's pluralization rules. English is one of the simpler cases: a subkey of "one" for the value 1, and a subkey of "other" for all other values. Since the value of `:count` is 33, the "utils.minutes.other" message will be used ("%{count} minutes"); then the value will be interpolated, and the result will be "33 minutes".

This is why I say it's not as simple as sending the arguments of `t` straight to the console. :smile_cat:

But like I said, a shim that does only what I describe above just to get the English text output should be feasible, if it's really necessary.

**Comment by mikemcquaid** [comment]
> But like I said, a shim that does only what I describe above just to get the English text output should be feasible, if it's really necessary.

At the moment we still have to support 10.8 for Mountain Lion so I think this might be tricky.

I wonder either if we roll our own solution based on a similar approach, find another gem that does similar and support 1.8 or fix up this gem to work on 1.8.

File: Library/Homebrew/vendor/i18n.rb (hunk 3 above)
Comment (mikemcquaid) [inline Library/Homebrew/vendor/i18n.rb]: Oh, nothing in particular, just feels like a lot of configuration.

File: .gitignore (hunk 1 above)
Comment (mikemcquaid) [inline .gitignore]: Yeh, thanks. Or Library/Homebrew/i18n?

**Comment by L2G** [comment]
> At the moment we still have to support 10.8 for Mountain Lion so I think this might be tricky.

What is the tricky part?

I've already rolled i18n back to 0.6.11 specifically to address the concern with Ruby 1.8 compatibility. 

You're worried about compatibility with Mountain Lion, but @mistydemeo has already told me she's worried about compatibility with Tiger.

**Comment by mikemcquaid** [comment]
> What is the tricky part?
> 
> I've already rolled i18n back to 0.6.11 specifically to address the concern with Ruby 1.8 compatibility.
> 
> You're worried about compatibility with Mountain Lion, but @mistydemeo has already told me she's worried about compatibility with Tiger.

I think we might be better trying to fixup the latest version rather than sticking on an old version but ultimately it's up to you there. I don't think we need to necessarily let Tigerbrew block this being included into Homebrew.

**Comment by L2G** [comment]
So… where have we landed on this? Use the latest i18n gem, but since it doesn't officially support Ruby 1.8.x, backport it so that it does?

**Comment by L2G** [comment]
By the way, if anyone knows of a foolproof way to build Ruby 1.8.7-p374 on Mavericks with ruby-install, I'd be grateful… :sweat:

**Comment by mistydemeo** [comment]
> I don't think we need to necessarily let Tigerbrew block this being included into Homebrew.

I don't want to block this either - I just want to make sure that it doesn't get so entwined Homebrew code that I can't opt out of it in Tigerbrew, in the event that it can't be made to work on old Rubies.

> By the way, if anyone knows of a foolproof way to build Ruby 1.8.7-p374 on Mavericks with ruby-install, I'd be grateful…

There's a `ruby187` formula in `homebrew/versions`! Not ruby-install, but...

When I'm not on my old Macs, I use the `ruby182`, `ruby186` and `ruby187` formulae to test things out.

**Comment by L2G** [comment]
Ohhhh, the homebrew/versions tap. \*forehead smack\*

**Comment by L2G** [comment]
Okay. I installed ruby182 and ruby187. For each of them, I installed Bundler, then in my clone of the i18n repo, I ran `bundle install` to install dependencies, and I ran `bundle exec rake test` to see what I could shake loose.

All tests passed, both versions of Ruby. So maybe we're in better shape than we all feared.

**Comment by jacknagel** [comment]
Can we fork the gem to homebrew/i18n, make our changes, run the tests, and then merge only what we need into the vendor directory? I'd really like to avoid having the i18n test suite in this repository.

**Comment by L2G** [comment]
@jacknagel Watch for my next push, because I have it stripped down to just the contents of the lib/ subdirectory of the i18n gem.

I'm also adding some tests in Library/Homebrew/test that will validate basic i18n operation.

**Comment by mikemcquaid** [comment]
> Can we fork the gem to homebrew/i18n, make our changes, run the tests, and then merge only what we need into the vendor directory? I'd really like to avoid having the i18n test suite in this repository.

This sounds great. We could also set up Travis builds for older Ruby versions in that repo too.

**Comment by mikemcquaid** [comment]
@L2G I've added an i18n fork to the Homebrew organisation and invited you to the i18n team. When you accept you should have push permissions there.

**Comment by L2G** [comment]
@mikemcquaid Thanks. I'm not clear on what I need to do, though. I haven't received any emails or prompts asking me to accept new permissions or membership for anything.

**Comment by L2G** [comment]
@mikemcquaid Oops. Seems as if GitHub made me hunt for it, but I got it.

**Comment by mistydemeo** [comment]
> Okay. I installed ruby182 and ruby187. For each of them, I installed Bundler, then in my clone of the i18n repo, I ran bundle install to install dependencies, and I ran bundle exec rake test to see what I could shake loose.

Fantastic!

What version of Bundler did you install? I had trouble finding one that was compatible with ruby182, partly because it doesn't include Rubygems.

**Comment by L2G** [comment]
@mistydemeo Maybe I did it wrong. :grimacing: 

    $ chruby
     * ruby-1.8.2
       ruby-1.8.7-p374
       ruby-2.1.5
       ruby-2.2.1

    $ which ruby
    /Users/larry/.rubies/ruby-1.8.2/bin/ruby

    $ which gem
    /usr/bin/gem
    
    $ which bundle
    /Users/larry/.gem/ruby/1.8.2/bin/bundle

**Comment by mistydemeo** [comment]
Check the shebang of `/Users/larry/.gem/ruby/1.8.2/bin/bundle`? It may have actually been installed using your system ruby.

**Comment by L2G** [comment]
Yeah, that's what I was afraid of as soon as you said that.

And...

    $ head /Users/larry/.gem/ruby/1.8.2/bin/bundle
    #!/System/Library/Frameworks/Ruby.framework/Versions/2.0/usr/bin/ruby

Poo. Will have to work on this some more.

**Comment by L2G** [comment]
@mistydemeo You said you had trouble finding a version of _Bundler_ that was compatible with Ruby 1.8.2. Have you gotten _RubyGems itself_ to work? The newest version I can find that works is RubyGems 0.9.5, but so far I've had no luck having it use https://rubygems.org as a source.

**Comment by mikemcquaid** [comment]
You may need to do stuff like disable SSL, from my vague memory.

**Comment by L2G** [comment]
I'm going to put it on the back burner for now.

**Comment by tdsmith** [comment]
@L2G Should we leave this pull request open?

**Comment by mikemcquaid** [comment]
I think this is probably conflicted enough it's worth reopening in future after we've moved to Ruby 2.0 for all of Homebrew (which should be some time this year).

**Pull request closed without merging**